source/BreachNode.cpp

```diff
@@ -54,7 +62,9 @@ bool BreachNode::init(std::shared_ptr<BreachModel> breach, std::shared_ptr<Donut
 	return true;
 }
 
-bool BreachNode::isActive() { return breachModel->getHealth() > 0 || isAnimatingShrink; }
+bool BreachNode::isActive() {
+	return breachModel->getHealth() > 0 || (prevHealth == 1 && breachModel->getHealth() == 0);
```

Comment:
This line should really just return `breachModel->getIsActive()`. If `isActive` becomes false before you want to scene graph node to despawn, then you should edit the game logic so that `isActive` stays true until the scene graph node can be safely despawned. It's not safe to leave scene graph nodes hanging around on screen after the model becomes inactive, because then the linked model becomes eligible to be recycled. This is how we get bugs like the -1 on the buttons.

Reply:
I did not like messing with the `isActive` method, but you have to know that the architecture is INCREDIBLY unaccommodating for animations because the nodes are written to be strictly responsive to models, which leaves very little opportunity for animation checks to run. For many of the in-game animations I've coded, I've almost always had to wrestle with the architecture because models get recycled, so it's basically impossible to implement trailing animations like the breach disappearing from health 1 into nothing. 

Here I need a one-frame window after the breach is resolved to activate the sparkle animation. However once the breach is resolved, the model and the node become inactive and `postPosition()` is no longer called. I could put some of the logic in `draw()`, but that would mean duplicate if statements and code, which kinda defeats the purpose of the `CustomNode` pipeline. Here I'm letting `isActive` be true for one extra frame after breach resolve to let `postPosition()` be executed. 

When we were discussing the `CustomNode` refactor, I said I was concerned about future code flexibilty for new features and this is the exact problem.

Reply:
Doors also have to stay on screen after being resolved, and exhibits the way this architecture was designed to work with this issue. If the information is needed after the resolution of the object, the model should not report `isActive` false. As soon as it does, then the model might be reused for another spawned object, which would break the scene nodes. If you need the model to stick around for the duration of the animation (which is probably a good idea here), that logic should probably go inside the model object itself.

Reply:
In the case of the door I agree that it's justified to have the door height be information that is a part of the model because it affects collision, so the model can set its `isActive` field in consideration of the height. 

However, with breaches, the trailing animation has no significance to the model itself.  Once the breach's health hits zero, it no longer impacts any other models and should be for all purposes considered inactive. However, on the display side, we want a trailing animation that is purely visual effects and does not reflect the current model state. The architecture simply does not accommodate trailing animations.

Reply:
Would you prefer if `CustomNode` exposed an additional lifecycle method that is called once upon a state change from active to not active? Would that make life easier?

Reply:
Yes that would resolve the issue of the `isActive()` changes. However it would also mean duplicate code for intiating the sparkle animation.

Reply:
The code to initiate the animation could always be factored into a different method inside `BreachNode`, which is then called in two places. I'll branch off this branch and work on adding that extra method. I'll ping you on Slack when I'm done.